pkg/resource/plugin/provider.go

```diff
@@ -41,7 +42,7 @@ type Provider interface {
 	// CheckConfig validates the configuration for this resource provider.
 	CheckConfig(olds, news resource.PropertyMap) (resource.PropertyMap, []CheckFailure, error)
 	// DiffConfig checks what impacts a hypothetical change to this provider's configuration will have on the provider.
-	DiffConfig(olds, news resource.PropertyMap) (DiffResult, error)
+	DiffConfig(olds, news resource.PropertyMap) (DiffResult, result.Result)
```

Comment:
I admit, I don't like this - this feels like a layering violation to me. I'm not a fan of mixing and matching result.Result and error as return types for different methods.

I think that we should consider logging everything that comes out of `Diff` and `DiffConfig` and keep it returning an `error`, to keep our code consistent here.

Reply:
I don't think it's wrong to have sentinel errors like `DiffUnknown` coming out of these functions - at least that way a caller can inspect the error and, if it's not a sentinel error, log a diagnostic and return a Bail.

Reply:
I totally agree with this feedback.